kvbc/tools/migrations/v4migration_tool/src/migration_bookeeper.cpp

```diff
@@ -0,0 +1,262 @@
+  max_number_of_batches_ = (last_block_id_ - start_block_id_) / point_lookup_batch_size;
+  if (((last_block_id_ - start_block_id_) % point_lookup_batch_size) != 0) {
+    max_number_of_batches_++;
+  }
+
+  for (uint64_t block_id = start_block_id_; block_id <= last_block_id_;) {
+    auto last_block_id_for_this_batch =
+        ((block_id + point_lookup_batch_size) <= last_block_id_) ? block_id + point_lookup_batch_size : last_block_id_;
+    tasks.push_back(batch_thread_pool_->async(
+        [this, &from](auto l_batch_id, auto l_from_block_id, auto l_last_block_id) {
+          read_from(from, l_batch_id, l_from_block_id, l_last_block_id);
+        },
+        max_read_batch_id_.load(),
+        block_id,
+        last_block_id_for_this_batch));
+    if (tasks.size() >= number_of_batches_allowed_in_cycle_) {
+      for (auto const& t : tasks) {
+        t.wait();
+      }
+      tasks.clear();
+      if (max_read_batch_id_.load() > max_write_batch_id_.load()) {
+        std::unique_lock<std::mutex> lock(read_batch_mutex_);
+        bool wait_for_batch_size_reduction = true;
+        while (wait_for_batch_size_reduction) {
+          wait_for_batch_size_reduction = !(read_batch_cond_var_.wait_for(lock, std::chrono::seconds(2), [this]() {
+            return max_read_batch_id_.load() <= max_write_batch_id_.load();
+          }));
+        }
+      }
+    }
+    ++max_read_batch_id_;
+    block_id = last_block_id_for_this_batch + 1;
+  }
+
+  for (auto const& t : tasks) {
+    t.wait();
+  }
+  read_completed_.store(true);
+  if (storage_thread_->joinable()) {
+    storage_thread_->join();
+  }
+}
+
+void BookKeeper::read_from(const std::shared_ptr<concord::kvbc::adapter::ReplicaBlockchain>& from,
+                           uint64_t batch_id,
+                           uint64_t from_block_id,
+                           uint64_t last_block_id) {
+  auto batch = std::make_shared<std::vector<concord::kvbc::categorization::Updates>>();
+  for (uint64_t block_id = from_block_id; block_id <= last_block_id; ++block_id) {
+    batch->push_back((from->getBlockUpdates(block_id)).value());
+  }
+  {
+    std::unique_lock<std::mutex> lock(batch_mutex_);
+    ConcordAssertNE(single_batch_, nullptr);
+    single_batch_->emplace(batch_id, batch);
+    batch_cond_var_.notify_all();
+  }
+}
+
+void BookKeeper::store_to(std::shared_ptr<concord::kvbc::adapter::ReplicaBlockchain>& to) {
+  max_write_batch_id_ = 1;
+  int completed_percent = -1;
+  while (true) {
+    bool wait_pred_result = false;
+    {
+      std::unique_lock<std::mutex> lock(batch_mutex_);
+      auto max_batch_id = max_read_batch_id_.load();
+      if ((max_batch_id != 0) && read_completed_.load() && (max_write_batch_id_.load() >= max_batch_id)) {
+        return;
+      }
+      wait_pred_result = batch_cond_var_.wait_for(
+          lock, std::chrono::seconds(1), [this]() { return max_write_batch_id_.load() <= max_read_batch_id_.load(); });
```

Comment:
the previous check for termination and this predicate satisfy if ```max_write_batch_id_== max_batch_id``` 
and since updating max_read_batch_id_ and read_completed_ is not atomic, can we have a race where max_write_batch_id_ == max_batch_id and read_completed_ is false though in practice is about to be true yet this thread performs those checks before the other sets it to true?

Reply:
Thanks @cloudnoize , this is a good catch. The termination condition should not have equal to in it.